### DISCUSSION TITLE
请问进行登陆之后有获得已收藏本子的方法吗？

### DISCUSSION THREAD

**Discussion by FengZhe-ZYH** [hect0x7/JMComic-Crawler-Python#235] [Q&A] [answered]
请问有模块能帮助获得账号下已收藏本子的id序列吗？想通过这个来对收藏的内容进行批量下载

**Comment by hect0x7** [comment, marked as answer]
v2.4.1 已实现，参考如下代码：

```python
from jmcomic import *

op = create_option('xxx')
cl = op.new_jm_client()

# 登录，可以使用插件或配置AVS等其他方式
cl.login('username', 'password')

# 遍历收藏页
for page in cl.favorite_folder_gen(username='username'): # 指定username参数，api Client 可以不传
    # 遍历收藏页下的每个结果
    for aid, atitle in page:
        # aid: 本子的album_id
        print(aid)
```

**Comment by FengZhe-ZYH** [comment]
> v2.4.1 已实现，参考如下代码：
> 
> ```python
> from jmcomic import *
> 
> op = create_option('xxx')
> cl = op.new_jm_client()
> 
> # 登录，可以使用插件或配置AVS等其他方式
> cl.login('username', 'password')
> 
> # 遍历收藏页
> for page in cl.favorite_folder_gen(username='username'): # 指定username参数，api Client 可以不传
>     # 遍历收藏页下的每个结果
>     for aid, atitle in page:
>         # aid: 本子的album_id
>         print(aid)
> ```

非常感谢您的工作，另外请问cl.favorite_folder(username='xxx')这个获取收藏夹信息功能是在完善中吗？我无法通过它获得我的文件夹信息。当然也有可能是我使用错误

**Comment by hect0x7** [comment]
> > v2.4.1 已实现，参考如下代码：
> > ```python
> > from jmcomic import *
> > 
> > op = create_option('xxx')
> > cl = op.new_jm_client()
> > 
> > # 登录，可以使用插件或配置AVS等其他方式
> > cl.login('username', 'password')
> > 
> > # 遍历收藏页
> > for page in cl.favorite_folder_gen(username='username'): # 指定username参数，api Client 可以不传
> >     # 遍历收藏页下的每个结果
> >     for aid, atitle in page:
> >         # aid: 本子的album_id
> >         print(aid)
> > ```
> 
> 非常感谢您的工作，另外请问cl.favorite_folder(username='xxx')这个获取收藏夹信息功能是在完善中吗？我无法通过它获得我的文件夹信息。当然也有可能是我使用错误

这个方法是完全可用的。
可能是你使用有误，你可以贴一下你的代码和option配置吗（记得脱敏）？

**Comment by hect0x7** [comment]
贴一个favorite_folder最简单的使用方式，无option配置，直接运行代码即可。

```python
from jmcomic import JmOption, JmApiClient

cl = JmOption.default().new_jm_client(impl=JmApiClient)
cl.login('你的用户名', '你的密码')
# 获取收藏夹 （第一页）
for aid, atitle in cl.favorite_folder():
    print(aid)

```

**Comment by FengZhe-ZYH** [comment]
> > > v2.4.1 已实现，参考如下代码：
> > > ```python
> > > from jmcomic import *
> > > 
> > > op = create_option('xxx')
> > > cl = op.new_jm_client()
> > > 
> > > # 登录，可以使用插件或配置AVS等其他方式
> > > cl.login('username', 'password')
> > > 
> > > # 遍历收藏页
> > > for page in cl.favorite_folder_gen(username='username'): # 指定username参数，api Client 可以不传
> > >     # 遍历收藏页下的每个结果
> > >     for aid, atitle in page:
> > >         # aid: 本子的album_id
> > >         print(aid)
> > > ```
> > 
> > 
> > 非常感谢您的工作，另外请问cl.favorite_folder(username='xxx')这个获取收藏夹信息功能是在完善中吗？我无法通过它获得我的文件夹信息。当然也有可能是我使用错误
> 
> 这个方法是完全可用的。 可能是你使用有误，你可以贴一下你的代码和option配置吗（记得脱敏）？

是的，这个方法完全能够正确地返回atittle以及aid，但是当我尝试传递参数folder = 'folder_name'进去时它仍然会返回所有的收藏内容而不是特定收藏夹下的内容

**Comment by hect0x7** [comment]
> > > > v2.4.1 已实现，参考如下代码：
> > > > ```python
> > > > from jmcomic import *
> > > > 
> > > > op = create_option('xxx')
> > > > cl = op.new_jm_client()
> > > > 
> > > > # 登录，可以使用插件或配置AVS等其他方式
> > > > cl.login('username', 'password')
> > > > 
> > > > # 遍历收藏页
> > > > for page in cl.favorite_folder_gen(username='username'): # 指定username参数，api Client 可以不传
> > > >     # 遍历收藏页下的每个结果
> > > >     for aid, atitle in page:
> > > >         # aid: 本子的album_id
> > > >         print(aid)
> > > > ```
> > > 
> > > 
> > > 非常感谢您的工作，另外请问cl.favorite_folder(username='xxx')这个获取收藏夹信息功能是在完善中吗？我无法通过它获得我的文件夹信息。当然也有可能是我使用错误
> > 
> > 
> > 这个方法是完全可用的。 可能是你使用有误，你可以贴一下你的代码和option配置吗（记得脱敏）？
> 
> 是的，这个方法完全能够正确地返回atittle以及aid，但是当我尝试传递参数folder = 'folder_name'进去时它仍然会返回所有的收藏内容而不是特定收藏夹下的内容

参数传得不对，特定收藏夹，对应的参数是folder_id，文件夹id。而不是folder_name文件夹名。
你可以参考禁漫网页url：
![image](https://github.com/hect0x7/JMComic-Crawler-Python/assets/93357912/38803af0-e028-4dd2-969d-8ef026b8e900)

```python
cl.favorite_folder(folder_id='18xxxx')
```

**Comment by hect0x7** [comment]
你也可以使用代码来获取目标文件夹的id
```python
from jmcomic import *

cl = JmOption.default().new_jm_client(impl=JmApiClient)
cl.login('xxx', 'yyy')
page = cl.favorite_folder()
for folder_info in page.folder_list:
    fname, fid = folder_info['name'], folder_info['FID']
    print(f'文件夹名: {fname}，文件夹id: {fid}')
```

**Comment by FengZhe-ZYH** [comment]
> 你也可以使用代码来获取目标文件夹的id
> 
> ```python
> from jmcomic import *
> 
> cl = JmOption.default().new_jm_client(impl=JmApiClient)
> cl.login('xxx', 'yyy')
> page = cl.favorite_folder()
> for folder_info in page.folder_list:
>     fname, fid = folder_info['name'], folder_info['FID']
>     print(f'文件夹名: {fname}，文件夹id: {fid}')
> ```


![image](https://github.com/hect0x7/JMComic-Crawler-Python/assets/121104554/a8347242-8667-438e-88b4-afa256a05b40)
![image](https://github.com/hect0x7/JMComic-Crawler-Python/assets/121104554/d6a0715c-c836-4bd4-af18-45ca9cd5d7e1)
感谢你的指导，但是当我传入参数如上图所示时仍然返回了所有的收藏内容，并且我注意到此时page.folder_list是一个空的列表，这合理吗？

**Comment by hect0x7** [comment]
> > 你也可以使用代码来获取目标文件夹的id
> > ```python
> > from jmcomic import *
> > 
> > cl = JmOption.default().new_jm_client(impl=JmApiClient)
> > cl.login('xxx', 'yyy')
> > page = cl.favorite_folder()
> > for folder_info in page.folder_list:
> >     fname, fid = folder_info['name'], folder_info['FID']
> >     print(f'文件夹名: {fname}，文件夹id: {fid}')
> > ```
> 
> ![image](https://user-images.githubusercontent.com/121104554/284276725-a8347242-8667-438e-88b4-afa256a05b40.png) ![image](https://user-images.githubusercontent.com/121104554/284276891-d6a0715c-c836-4bd4-af18-45ca9cd5d7e1.png) 感谢你的指导，但是当我传入参数如上图所示时仍然返回了所有的收藏内容，并且我注意到此时page.folder_list是一个空的列表，这合理吗？

注意 new_jm_client(impl=JmApiClient)，需要指定impl参数为JmApiClient，也就是使用移动端api。
原因是目前网页端的Client还没完整实现page.folder_list
你使用了配置文件，那就改一下new_option.yml就可以了，代码不用改：

```yml
client:
  impl: api
```

**Comment by FengZhe-ZYH** [comment]
十分感谢您迅速的回复，完全解决了我的问题。
这里贴出一个可能会遇到的包导入错误： ModuleNotFoundError: No module named 'Crypto'
当使用impl: api时可能会产生这个错误，这个时候请检查pycryptodome与Crypto的两个库是否包含，安装顺序相反也可能会导致此错误
~~~
pip install pycryptodome
pip install Crypto
~~~

**Comment by hect0x7** [comment]
> 十分感谢您迅速的回复，完全解决了我的问题。 这里贴出一个可能会遇到的包导入错误： ModuleNotFoundError: No module named 'Crypto' 当使用impl: api时可能会产生这个错误，这个时候请检查pycryptodome与Crypto的两个库是否包含，安装顺序相反也可能会导致此错误
> 
> ```
> pip install pycryptodome
> pip install Crypto
> ```

噢是的，问题在于jmcomic的setup漏了这个库，导致在安装jmcomic时没有自动装上这个库依赖，感谢你的补充
